synapse/rest/client/register.py

```diff
@@ -869,6 +869,65 @@ async def _do_guest_registration(
         return 200, result
 
 
+class RegisterAppServiceOnlyRestServlet(RestServlet):
+    """An alternative registration API endpoint that only allows ASes to register"""
+
+    PATTERNS = client_patterns("/register$")
+    CATEGORY = "Registration/login requests"
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__()
+
+        self.auth = hs.get_auth()
+        self.registration_handler = hs.get_registration_handler()
+        self.ratelimiter = hs.get_registration_ratelimiter()
+
+    @interactive_auth_handler
+    async def on_POST(self, request: SynapseRequest) -> Tuple[int, JsonDict]:
+        body = parse_json_object_from_request(request)
+
+        client_addr = request.getClientAddress().host
+
+        await self.ratelimiter.ratelimit(None, client_addr, update=False)
+
+        kind = parse_string(request, "kind", default="user")
+
+        if kind == "guest":
+            raise SynapseError(403, "Guest access is disabled")
+        elif kind != "user":
+            raise UnrecognizedRequestError(
+                f"Do not understand membership kind: {kind}",
+            )
+
+        # Pull out the provided username and do basic sanity checks early since
+        # the auth layer will store these in sessions.
+        desired_username = body.get("username")
+        if not isinstance(desired_username, str) or len(desired_username) > 512:
+            raise SynapseError(400, "Invalid username")
+
+        # Allow only ASes to use this API.
+        if body.get("type") != APP_SERVICE_REGISTRATION_TYPE:
+            raise SynapseError(403, "Non-application service registration type")
+
+        if not self.auth.has_access_token(request):
+            raise SynapseError(
+                400,
+                "Appservice token must be provided when using a type of m.login.application_service",
+            )
+
+        # XXX we should check that desired_username is valid. Currently
+        # we give appservices carte blanche for any insanity in mxids,
+        # because the IRC bridges rely on being able to register stupid
+        # IDs.
+
+        as_token = self.auth.get_access_token_from_request(request)
+
+        user_id = await self.registration_handler.appservice_register(
+            desired_username, as_token
+        )
```

Comment:
This seems to skip the whole `_create_registration_details` flow?

Reply:
Yeah, this endpoint no longer return an access token. Appservices weren't really using that anyway

Reply:
But they *could* use it, right? I think there are appservices which use access tokens. Can you expand a bit about how they aren't using it?

Reply:
Appservices call the C-S API either by:

 - using a "real" user-specific access token, by doing a `m.login.application_service` on /login or /register
 - masquerading using the `user_id` query parameter and their AS token

The first one doesn't work with MSC3861 for now, and we don't have an OIDC-native alternative for now.
Non-E2EE appservices don't usually need it, since the only advantage of getting a real access token is to have a real device for AS users.
MSC3202 adds device masquerading for ASes, so matrix-bot-sdk (used by Hookshot) will use that if it is available and fallback to /login if not. So, as long as MSC3202 is enabled, most modern appservices should work.

I expended the docstring on this handler here cb8a560e7

Reply:
OK, so I think MSC3861 has a soft-dependency on MSC3202 then.

Reply:
Not exactly? It's more that MSC3861 doesn't cover ASes yet, and a workaround for now is to not break ASes the least we can, and enabling MSC3202 is enough to make most modern ASes work.

Reply:
I added a thread for tracking: https://github.com/matrix-org/matrix-spec-proposals/pull/3861/files#r1206925965